### PR TITLE
[Modules] Added multiple vm deployment example

### PR DESCRIPTION
**Pull request by MrMCake**
# Description

- Added a sample/construct to show how to deploy multiple VMs using our module in combination with a name generator
- Tested both conditional `vmNames` and `vmInitialNumber` + `vmNamePrefix` + `vmNumberOfInstances`

vnamesportal

# Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

# Example output of deployment

```PowerShell

DeploymentName          : vmNamesTest
ResourceGroupName       : validation-rg
ProvisioningState       : Succeeded
Timestamp               : 06.05.2022 20:03:21
Mode                    : Incremental
TemplateLink            : 
Parameters              : 
                          Name                                     Type                       Value
                          =======================================  =========================  ==========
                          vmNames                                  Array                      ["alsehr-vm-linux-min-01","alsehr-vm-linux-min-02","alsehr-vm-linux-min-03"]
                          vmNamePrefix                             String                     "alsehr"
                          vmInitialNumber                          Int                        111
                          vmNumberOfInstances                      Int                        3
                          vmComputerNamesTransformation            String                     "none"
                          vmSize                                   String                     "Standard_B12ms"
                          encryptionAtHost                         Bool                       true
                          securityType                             String                     ""
                          secureBootEnabled                        Bool                       false
                          vTpmEnabled                              Bool                       false
                          imageReference                           Object                     {"publisher":"Canonical","offer":"UbuntuServer","sku":"18.04-LTS","version":"latest"}
                          plan                                     Object                     {}
                          osDisk                                   Object                     {"diskSizeGB":"128","managedDisk":{"storageAccountType":"Premium_LRS"}}
                          dataDisks                                Array                      []
                          ultraSSDEnabled                          Bool                       false
                          adminUsername                            SecureString               null
                          adminPassword                            SecureString               null
                          customData                               String                     ""
                          certificatesToBeInstalled                Array                      []
                          vmPriority                               String                     "Regular"
                          enableEvictionPolicy                     Bool                       false
                          maxPriceForLowPriorityVm                 String                     ""
                          dedicatedHostId                          String                     ""
                          licenseType                              String                     ""
                          publicKeys                               Array                      [{"path":"/home/localAdminUser/.ssh/authorized_keys","keyData":"..."}]
                          systemAssignedIdentity                   Bool                       false
                          userAssignedIdentities                   Object                     {}
                          bootDiagnosticStorageAccountName         String                     ""
                          bootDiagnosticStorageAccountUri          String                     ".blob.core.windows.net/"
                          proximityPlacementGroupName              String                     ""
                          availabilitySetName                      String                     ""
                          availabilityZone                         Int                        0
                          nicConfigurations                        Array                      [{"nicSuffix":"-nic-01","ipConfigurations":[{"name":"ipconfig01","subnetId":"/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Network/virtualNetworks/adp-<<namePrefix>>-az-vnet-x-001/subnets/<<namePrefix>>-az-subnet-x-001","pipC 
                          onfiguration":{"publicIpNameSuffix":"-pip-01"}}]}]
                          pipDiagnosticSettingsName                String                     ""        
                          pipdiagnosticLogCategoriesToEnable       Array                      ["DDoSProtectionNotifications","DDoSMitigationFlowLogs","DDoSMitigationReports"]
                          pipdiagnosticMetricsToEnable             Array                      ["AllMetrics"]
                          nicDiagnosticSettingsName                String                     ""
                          nicdiagnosticMetricsToEnable             Array                      ["AllMetrics"]
                          backupVaultName                          String                     ""
                          backupVaultResourceGroup                 String                     "validation-rg"
                          backupPolicyName                         String                     "DefaultPolicy"
                          enableServerSideEncryption               Bool                       false
                          allowExtensionOperations                 Bool                       true
                          extensionDomainJoinPassword              SecureString               null
                          extensionDomainJoinConfig                Object                     {"enabled":false}
                          extensionAntiMalwareConfig               Object                     {"enabled":false}
                          extensionMonitoringAgentConfig           Object                     {"enabled":false}
                          monitoringWorkspaceId                    String                     ""
                          extensionDependencyAgentConfig           Object                     {"enabled":false}
                          extensionNetworkWatcherAgentConfig       Object                     {"enabled":false}
                          extensionDiskEncryptionConfig            Object                     {"enabled":false}
                          extensionDSCConfig                       Object                     {"enabled":false}
                          extensionCustomScriptConfig              Object                     {"enabled":false,"fileData":[]}
                          extensionCustomScriptProtectedSetting    SecureObject               null
                          location                                 String                     "westeurope"
                          diagnosticLogsRetentionInDays            Int                        365
                          diagnosticStorageAccountId               String                     ""
                          diagnosticWorkspaceId                    String                     ""
                          diagnosticEventHubAuthorizationRuleId    String                     ""
                          diagnosticEventHubName                   String                     ""
                          lock                                     String                     "NotSpecified"
                          roleAssignments                          Array                      []
                          tags                                     Object                     {}
                          enableDefaultTelemetry                   Bool                       true
                          baseTime                                 String                     "2022-05-06 20:01:58Z"
                          sasTokenValidityLength                   String                     "PT8H"
                          osType                                   String                     "Linux"
                          disablePasswordAuthentication            Bool                       true
                          provisionVMAgent                         Bool                       true
                          enableAutomaticUpdates                   Bool                       true
                          timeZone                                 String                     ""
                          additionalUnattendContent                Array                      []
                          winRM                                    Object                     {}
                          configurationProfileAssignments          Array                      []

Outputs                 : 
                          Name                            Type                       Value
                          ==============================  =========================  ==========
                          virtualMachinesResourceId       Array                      ["/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Com 
                          pute/virtualMachines/alsehr-vm-linux-min-01","/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Compute/virtualMac 
                          hines/alsehr-vm-linux-min-02","/subscriptions/<<subscriptionId>>/resourceGroups/validation-rg/providers/Microsoft.Compute/virtualMachines/alsehr-vm 
                          -linux-min-03"]
                          virtualMachinesResourceGroup    String                     "validation-rg"
                          virtualMachinesName             Array                      ["alsehr-vm-linux-min-01","alsehr-vm-linux-min-02","alsehr-vm-linux-min-03"]

DeploymentDebugLogLevel : 
```

### CODE REVIEW

File: constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json [hunk 1]
```diff
@@ -0,0 +1,67 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "vmInitialNumber": {
+            "value": 111
+        },
+        "vmNamePrefix": {
+            "value": "alsehr"
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json:9]: So autoreferential

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+## Outputs
+
+| Output Name | Type | Description |
+| :-- | :-- | :-- |
+| `virtualMachinesName` | array | The Names of the VMs. |
+| `virtualMachinesResourceGroup` | string | The name of the Resource Group the VM(s) was/were created in. |
+| `virtualMachinesResourceId` | array | The Resource Id(s) of the VM(s). |
+
+## Template references
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md:197]: Remove

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+},
+```
+
+## Outputs
+
+| Output Name | Type | Description |
+| :-- | :-- | :-- |
+| `virtualMachinesName` | array | The Names of the VMs. |
+| `virtualMachinesResourceGroup` | string | The name of the Resource Group the VM(s) was/were created in. |
+| `virtualMachinesResourceId` | array | The Resource Id(s) of the VM(s). |
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md:195]:
```suggestion
| `virtualMachinesResourceId` | array | The Resource ID(s) of the VM(s). |
```

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        "/subscriptions/12345678-1234-1234-1234-123456789012/resourcegroups/validation-rg/providers/Microsoft.ManagedIdentity/userAssignedIdentities/adp-sxx-az-msi-x-002": {}
+    }
+},
+```
+
+## Outputs
+
+| Output Name | Type | Description |
+| :-- | :-- | :-- |
+| `virtualMachinesName` | array | The Names of the VMs. |
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md:193]:
```suggestion
| `virtualMachinesName` | array | The names of the VMs. |
```

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+| `virtualMachinesResourceGroup` | string | The name of the Resource Group the VM(s) was/were created in. |
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md:194]:
```suggestion
| `virtualMachinesResourceGroup` | string | The name of the resource group the VM(s) was/were created in. |
```

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep [hunk 6]
```diff
@@ -3,6 +3,11 @@
 param name string = take(toLower(uniqueString(resourceGroup().name)), 10)
 
 @description('Optional. Specifies whether the computer names should be transformed. The transformation is performed on all computer names. Available transformations are \'none\' (Default), \'uppercase\' and \'lowercase\'.')
+@allowed([
```
Comment (eriqua) [inline arm/Microsoft.Compute/virtualMachines/deploy.bicep:6]: Wondering about the general value of this parameter

File: constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.vmnames.parameter.json [hunk 7]
```diff
@@ -0,0 +1,65 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "vmNames": {
+            "value": [
+                "alsehr-vm-linux-min-01",
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.vmnames.parameter.json:7]: Replace min with vmnames

File: constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json (hunk 1 above)
Comment (eriqua) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json:9]: Maybe something like `<<namePrefix>>-vm-linux-prefix`

File: constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json [hunk 8]
```diff
@@ -0,0 +1,67 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "vmInitialNumber": {
+            "value": 111
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json:6]:
```suggestion
            "value": 1
```

Comment (MrMCake) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json]: done

File: constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.vmnames.parameter.json (hunk 7 above)
Comment (MrMCake) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.vmnames.parameter.json]: done

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md (hunk 5 above)
Comment (MrMCake) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md]: done

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md (hunk 4 above)
Comment (MrMCake) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md]: done

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md (hunk 3 above)
Comment (MrMCake) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md]: done

File: constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md (hunk 2 above)
Comment (MrMCake) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/readme.md]: done

File: constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json (hunk 1 above)
Comment (MrMCake) [reply constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json]: done

File: arm/Microsoft.Compute/virtualMachines/deploy.bicep (hunk 6 above)
Comment (MrMCake) [reply arm/Microsoft.Compute/virtualMachines/deploy.bicep:6]: Fair point. But I wouldn't remove it in this PR I guess. Might have been a specific requirement for the original implementation of the module that should not be carried over. Anyhow something to cover separatly if you agree

**Review by eriqua — COMMENTED** [review]
Should we have costructs tested? What's your take on that?

File: constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json [hunk 9]
```diff
@@ -0,0 +1,67 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "vmInitialNumber": {
+            "value": 1
+        },
+        "vmNamePrefix": {
+            "value": "<<namePrefix>>vm"
```
Comment (eriqua) [inline constructs/Microsoft.Compute/virtualMachinesMultiple/.parameters/linux.prefix.parameter.json:9]:
```suggestion
            "value": "<<namePrefix>>-vm-linux-prefix"
```

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**